template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/ranger/RangerCloudStorageServiceConfigProvider.java

```diff
@@ -0,0 +1,53 @@
+package com.sequenceiq.cloudbreak.cmtemplate.configproviders.ranger;
+
+import static com.sequenceiq.cloudbreak.cmtemplate.configproviders.ConfigUtils.config;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import org.springframework.stereotype.Component;
+
+import com.cloudera.api.swagger.model.ApiClusterTemplateConfig;
+import com.sequenceiq.cloudbreak.cmtemplate.CmTemplateComponentConfigProvider;
+import com.sequenceiq.cloudbreak.cmtemplate.CmTemplateProcessor;
+import com.sequenceiq.cloudbreak.cmtemplate.configproviders.ConfigUtils;
+import com.sequenceiq.cloudbreak.template.TemplatePreparationObject;
+
+@Component
+public class RangerCloudStorageServiceConfigProvider implements CmTemplateComponentConfigProvider {
+
+    private static final String RANGER_HDFS_AUDIT_URL = "ranger_plugin_hdfs_audit_url";
+
+    @Override
+    public List<ApiClusterTemplateConfig> getServiceConfigs(TemplatePreparationObject templatePreparationObject) {
+        List serviceConfigs = new ArrayList(1);
+        getCloudStorageProperties(templatePreparationObject)
+                .filter(s -> !s.isEmpty())
+                .ifPresent(storage -> serviceConfigs.add(config(RANGER_HDFS_AUDIT_URL, storage)));
+        return serviceConfigs;
+    }
+
+    @Override
+    public String getServiceType() {
+        return RangerRoles.RANGER;
+    }
+
+    @Override
+    public List<String> getRoleTypes() {
+        return List.of(RangerRoles.RANGER_ADMIN);
+    }
+
+    @Override
+    public boolean isConfigurationNeeded(CmTemplateProcessor cmTemplateProcessor, TemplatePreparationObject source) {
+        return source.getFileSystemConfigurationView().isPresent()
+                && cmTemplateProcessor.isRoleTypePresentInService(getServiceType(), getRoleTypes());
+    }
+
+    protected Optional<String> getCloudStorageProperties(TemplatePreparationObject source) {
+        StringBuilder rangerCloudStorage = new StringBuilder();
+        ConfigUtils.getStorageLocationForServiceProperty(source, RANGER_HDFS_AUDIT_URL).ifPresent(
```

Comment:
I think that `getStorageLocationForServiceProperty()` should have 2 variants: one can use the legacy service / Ambari property key (+ optionally a property file name), and the other would take a CM config ID tuple (service_name, role_name, template_parameter_key). This specific usage would require the second variant, as `getProperty()` will always reflect the legacy service / Ambari property key, never the template_parameter_key. (And template_parameter_key by itself is not guaranteed to be unique; multiple roles/services may use the same parameter name.)

Reply:
This code is in template-manager-cmtemplate module and is used only in CM wiring. Ambari Wiring is handled seperately in template-manager-blueprint module

Reply:
Yes, I know where it is. 😃 All I am saying is `RANGER_HDFS_AUDIT_URL` represents a CM template parameter key, which is something different than the underlying service config file (XML) property name. `getStorageLocationForServiceProperty()` is expecting the latter one, so it will not find any matches based on `RANGER_HDFS_AUDIT_URL`. Moreover, `StorageLocationView` will have to be extended with the corresponding CM template parameter key (and with possibly service_name and role_name) in order to make such lookups successful.

Reply:
RANGER_HDFS_AUDIT_URL is not a template variable. It is a new service config property introduced by Ranger Team to be used in CDP Ranger instead of xasecure.audit.destination.hdfs :). Changes to StorageLocationView is not part of this PR and is being worked on by CB team. Once that is done I will update this draft with the new code changes and create a new PR along with any property name changes. Also the CM template variable used and actual service variable can be different as used in HiveMetastoreCloudStorageRoleConfigProvider.
getStorageLocationForServiceProperty will always use the service property and not template property.